include/proxy-wasm/context.h

```diff
@@ -0,0 +1,306 @@
+  bool isVmContext() { return id_ == 0; }
+  // Root Contexts have the VM Context as a parent.
+  bool isRootContext() { return root_context_id_ == 0; }
+  ContextBase *root_context() { return root_context_; }
+  string_view root_id() const { return plugin_->root_id_; }
+  string_view log_prefix() const { return plugin_->log_prefix(); }
+  WasmVm *wasmVm() const;
+
+  /**
+   * Calls into the VM.
+   * These are implemented by the proxy-independent host code. They are virtual to support some
+   * types of testing.
+   */
+
+  // Context
+  void onCreate(uint32_t parent_context_id) override;
+  bool onDone() override;
+  void onDelete() override;
+
+  // Root
+  bool onStart(std::shared_ptr<PluginBase> plugin) override;
+  bool onConfigure(std::shared_ptr<PluginBase> plugin) override;
+  void onHttpCallResponse(HttpCallToken token) override;
+  void onQueueReady(uint32_t SharedQueueDequeueToken) override;
+  void onGrpcReceiveInitialMetadata(GrpcToken token) override;
+  void onGrpcReceive(GrpcToken token) override;
+  void onGrpcReceiveTrailingMetadata(GrpcToken token) override;
+  void onGrpcClose(GrpcToken token, const GrpcStatus &status, const string_view message) override;
+
+  // Stream (e.g. HTTP, Network)
+  void onFinalized() override;
+  const BufferInterface *getBuffer(WasmBufferType /* type */) override {
+    unimplemented();
+    return nullptr;
+  }
+  bool end_of_stream(WasmStreamType /* type */) override {
+    unimplemented();
+    return true;
+  }
+  WasmResult continueStream(WasmStreamType /* type */) override { return unimplemented(); }
+
+  // HTTP
+  ProxyAction onRequestHeaders() override;
+  ProxyAction onRequestBody() override;
+  ProxyAction onRequestTrailers() override;
+  ProxyAction onRequestMetadata() override;
+  ProxyAction onResponseHeaders() override;
+  ProxyAction onResponseBody() override;
+  ProxyAction onResponseTrailers() override;
+  ProxyAction onResponseMetadata() override;
+  WasmResult sendLocalResponse(uint64_t /* response_code */, string_view /* body */,
+                               Pairs /* additional_headers */, uint64_t /* grpc_status */,
+                               string_view /* details */) override {
+    return unimplemented();
+  }
+
+  // Network
+  ProxyAction onNetworkNewConnection() override;
+  ProxyAction onDownstreamData() override;
+  ProxyAction onUpstreamData() override;
+  using CloseType = NetworkInterface::CloseType;
+  void onDownstreamConnectionClose(CloseType close_type) override;
+  void onUpstreamConnectionClose(CloseType close_type) override;
+
+  // General
+  void error(string_view message) override {
+    std::cerr << message << "\n";
+    abort();
+  }
+  WasmResult unimplemented() override {
+    error("unimplemented proxy-wasm API");
+    return WasmResult::Unimplemented;
+  }
+  WasmResult log(uint64_t level, string_view message) override {
+    if (level >= static_cast<uint32_t>(LogLevel::error)) {
+      std::cerr << log_prefix() << message << "\n";
```

Comment:
seems like we are re-implementing the Envoy logging here.

Thinking more broadly, it looks like this implements an interface that I'm guessing you are trying to make independent of Envoy. OK fine.

But then I think you should have an Envoy-specific implementation of this class, which uses the Envoy infrastructure for logging, time, and probably lots of other things, rather than re-inventing them. WDYT?

I'm OK with a TODO here but I think we should get directionally aligned.

Reply:
log gets called in the worker thread, right? output to cerr might impact performance?

Reply:
The full Envoy version is already implemented in the envoy-wasm repo and in the upstreaming PR: https://github.com/envoyproxy/envoy/pull/10262

Reply:
This is overridden both by Envoy for production work and in tests for mocking purposes.  This is just a default implementation which integrators can use during the integration process.